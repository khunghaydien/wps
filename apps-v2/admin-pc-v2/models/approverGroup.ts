type Department = {
  code: string;
  name: string;
};

export type Approver = {
  id?: string;
  username?: string;
};

export type ApproverGroup = {
  id: string;
  companyId: string;
  code: string;
  department: Department;
  name: string;
  name_L0: string;
  name_L1: string;
  name_L2: string;
  queueId: string;
  attQueueId: string;
  expQueueId: string;
  generalQueueId: string;
  timeQueueId: string;
  type: keyof typeof TYPE_LABELS;
  targetDepartmentId: string;
  lastUpdatedAt: string;
};

export const TYPE_LABELS = {
  DeptManagers: 'Admin_Lbl_DepartmentManagers',
  DeptMembers: 'Admin_Lbl_DepartmentMembers',
  SpecificUsers: 'Admin_Lbl_SpecificUsers',
};

export const APPROVER_MODULES = {
  attQueueId: 'Admin_Lbl_EmpSectAttendance',
  expQueueId: 'Admin_Lbl_EmpSectExpense',
  timeQueueId: 'Admin_Lbl_EmpSectTimeTracking',
  generalQueueId: 'Admin_Lbl_CustomRequest',
};
