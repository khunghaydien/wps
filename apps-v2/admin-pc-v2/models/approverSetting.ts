export type ApproverSetting = {
  companyId: string;
  requestType: keyof typeof REQUEST_TYPE_LABELS;
  approver01: string;
  approver02: string;
  approver03: string;
  approver04: string;
  approver05: string;
};

export const COMPONENT_KEY = 'approverSetting';

export const DISABLE_REQUEST_TYPE_APPROVER01 = [
  'AttDailyRequest',
  'AttFixDailyRequest',
  'AttFixRequest',
  'AttAgreementRequest',
];

export const ENABLE_COST_CENTER_MANAGER_OPTION = [
  'ExpReportRequest',
  'ExpRequestApproval',
];

export const KEY_ENABLE_COST_CENTER_MANAGER_OPTION = 'CostCenterManager';

export const REQUEST_TYPE_LABELS = {
  AttDailyRequest: 'Appr_Lbl_AttendanceRequest',
  AttFixDailyRequest: 'Appr_Lbl_DailyFixRequest',
  AttFixRequest: 'Appr_Lbl_MonthlyFixRequest',
  AttAgreementRequest: 'Appr_Lbl_LegalAgreementRequest',
  TimeRequest: 'Appr_Btn_TimeTrackRequest',
  ExpReportRequest: 'Appr_Clbl_ExpensesRequest',
  ExpRequestApproval: 'Appr_Clbl_ExpensesPreApproval',
  ComGeneralRequest: 'Com_Tab_CustomRequest',
};

export const APPROVER_OPTION_LABELS = {
  EmpApprover01: 'Admin_Lbl_ApplicantApprover01',
  EmpApprover02: 'Admin_Lbl_ApplicantApprover02',
  EmpApprover03: 'Admin_Lbl_ApplicantApprover03',
  EmpApprover04: 'Admin_Lbl_ApplicantApprover04',
  EmpApprover05: 'Admin_Lbl_ApplicantApprover05',
  EmpApprover06: 'Admin_Lbl_ApplicantApprover06',
  EmpApprover07: 'Admin_Lbl_ApplicantApprover07',
  EmpApprover08: 'Admin_Lbl_ApplicantApprover08',
  EmpApprover09: 'Admin_Lbl_ApplicantApprover09',
  EmpApprover10: 'Admin_Lbl_ApplicantApprover10',
  EmpManager: 'Admin_Lbl_ApplicantManager',
  CostCenterManager: 'Admin_Lbl_RequestCostCenterManager',
  EmpPrimaryDeptMgr: 'Admin_Lbl_ApplicantPrimaryDeptManager',
  EmpPrimaryParentDeptMgr: 'Admin_Lbl_ApplicantPrimaryParentDeptManager',
  ReqJobOwner: 'Admin_Lbl_RequestJobOwner',
  ChoiceByApplicant: 'Admin_Lbl_ChoiceByApplicant',
  NotUsed: 'Admin_Lbl_NotUsed',
};
