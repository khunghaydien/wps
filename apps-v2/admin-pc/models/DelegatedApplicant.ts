import Api from '../../commons/api';

export type { EmployeeShowObj } from './DelegatedApprover';

export type DelegatedApplicant = {
  settingId: string | null;
  isActiveSFUserAcc?: boolean;
  delegateeId?: string;
  delegatee?: {
    id: string;
    name: string;
    photoUrl: string;
    department: {
      code: string;
      name: string;
    };
  };
  delegatedFor: {
    expense: boolean;
    request: boolean;
  };
};

export const getDelegatedApplicantList = (
  empBaseId: string
): Promise<DelegatedApplicant[]> => {
  return Api.invoke({
    path: '/delegated-application-setting/list',
    param: {
      empBaseId,
    },
  }).then((result: { settingList: Array<DelegatedApplicant> }) => {
    return result.settingList;
  });
};

export const saveDelegatedApplicants = (
  empId: string,
  settings: Array<DelegatedApplicant>
): Promise<void> => {
  return Api.invoke({
    path: '/delegated-application-setting/save',
    param: {
      empId,
      settings,
    },
  }).then((response) => response);
};
