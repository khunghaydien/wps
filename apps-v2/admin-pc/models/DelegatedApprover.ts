import Api from '../../commons/api';

export type EmployeeShowObj = {
  id: string;
  code: string;
  name: string;
  depCode: string;
  depName: string;
  photoUrl: string;
  title: string;
  isActiveSFUserAcc?: boolean;
};

export type DelegatedApprover = {
  settingId: string | null;
  delegatedApproverId: string;
  canApproveExpenseRequestByDelegate: boolean;
  canApproveExpenseReportByDelegate: boolean;
  delegatedApproverCode?: string | null | undefined;
  delegatedApproverName?: string | null | undefined;
  departmentCode?: string | null | undefined;
  departmentName?: string | null | undefined;
  delegatedApproverPhotoUrl?: string | null | undefined;
  isActiveSFUserAcc?: boolean;
};

export const getDelegatedApproverList = (
  empId: string
): Promise<DelegatedApprover[]> => {
  return Api.invoke({
    path: '/delegated-approver-setting/list',
    param: {
      empId,
    },
  }).then((result: { settingList: Array<DelegatedApprover> }) => {
    return result.settingList;
  });
};

export const saveDelegatedApprovers = (
  empId: string,
  settings: Array<DelegatedApprover>
): Promise<any> => {
  return Api.invoke({
    path: '/delegated-approver-setting/save',
    param: {
      empId,
      settings,
    },
  }).then((response) => response);
};
