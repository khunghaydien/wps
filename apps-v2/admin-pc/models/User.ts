import Api from '../../commons/api';

export const LIMIT_NUMBER = 100;

export type User = {
  userName: string;
  profileId: string;
  profile: {
    name: string;
  };
  name: string;
  id: string;
};

// eslint-disable-next-line import/prefer-default-export
export const search = (param: {
  name: string;
  userName: string;
}): Promise<User[]> => {
  return Api.invoke({
    path: '/user/search',
    param: {
      ...param,
      limitNumber: LIMIT_NUMBER + 1,
    },
  }).then((result: { records: Array<User> }) => {
    return result.records;
  });
};
