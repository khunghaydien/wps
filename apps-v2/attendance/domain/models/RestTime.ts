import { RestTimeReason } from './RestTimeReason';
import * as TimeRange from './TimeRange';

export type RestTime = TimeRange.TimeRange & {
  restReason: RestTimeReason;
};

export type RestTimes = RestTime[];

/**
 * 休憩時刻（開始時刻と終了時刻のペア）の上限
 */
export const MAX_STANDARD_REST_TIME_COUNT = 5;

export const create = <T extends RestTime>(
  restTime?: Partial<T>
): RestTime => ({
  startTime: restTime?.startTime ?? null,
  endTime: restTime?.endTime ?? null,
  restReason: restTime?.restReason ?? null,
});

export const convert = <
  T extends {
    startTime: any;
    endTime: any;
    restReason: RestTimeReason;
  }
>(
  restTime: T
): RestTime => ({
  ...TimeRange.convert(restTime),
  restReason: restTime.restReason,
});

export const hasValue = TimeRange.hasTimes;

export const convertRestTimes = (
  restTimes: {
    startTime: any;
    endTime: any;
    restReason: RestTimeReason;
  }[]
): RestTimes => restTimes.map(convert);

export const isLast =
  (idx: number) =>
  (records: RestTimes): boolean =>
    idx === records.length - 1;

export const isAddable =
  (maxLength = MAX_STANDARD_REST_TIME_COUNT) =>
  (idx: number) =>
  (_: RestTimes): boolean =>
    idx + 1 < maxLength;

export const isDeletable =
  (_: number) =>
  (records: RestTimes): boolean =>
    records.length > 1;

export type IRestTimeFactory<TRestTime extends RestTime = RestTime> = {
  create: (...args: unknown[]) => TRestTime;
};

export type IRestTimesFactory<TRestTime extends RestTime = RestTime> = {
  maxLength: number;
  create: () => TRestTime[];
  filter: (arr: TRestTime[]) => TRestTime[];
  update: (arr: TRestTime[], idx: number, value: TRestTime) => TRestTime[];
  insert: (arr: TRestTime[], idx: number, value?: TRestTime) => TRestTime[];
  remove: (arr: TRestTime[], idx: number) => TRestTime[];
  push: (arr: TRestTime[], value?: TRestTime) => TRestTime[];
  pushLast: (arr: TRestTime[]) => TRestTime[];
};
