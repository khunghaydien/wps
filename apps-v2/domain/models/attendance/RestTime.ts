import * as TimeRange from './TimeRange';

export type RestTimes = TimeRange.TimeRange[];

/**
 * 休憩時刻（開始時刻と終了時刻のペア）の上限
 * - TODO: 2017年11月末（Winter ’18）リリース時点での制限。いずれ無制限に対応するはず
 * @type {Number}
 */
export const MAX_STANDARD_REST_TIME_COUNT = 5;

export const create = (
  restTimes: {
    startTime: any;
    endTime: any;
  }[]
): RestTimes => restTimes.map(TimeRange.create);

export const filter = (restTimes: RestTimes): RestTimes =>
  restTimes.filter(TimeRange.hasTimes);

export const pushLast = (restTimes: RestTimes): RestTimes => {
  const arr = [...restTimes];
  if (
    arr.length < MAX_STANDARD_REST_TIME_COUNT &&
    (arr.length === 0 || TimeRange.hasTimes(arr[arr.length - 1]))
  ) {
    arr.push(TimeRange.create());
  }
  return arr;
};

export const update = (
  restTimes: RestTimes,
  idx: number,
  value: {
    startTime: any;
    endTime: any;
  }
): RestTimes => {
  const arr = [...restTimes];
  arr.splice(idx, 1, TimeRange.create(value));
  return arr;
};

export const remove = (restTimes: RestTimes, idx: number): RestTimes => {
  const arr = [...restTimes];
  arr.splice(idx, 1);
  return arr;
};

export const push = (
  restTimes: RestTimes,
  value: { startTime: any; endTime: any } = { startTime: null, endTime: null }
): RestTimes => {
  const arr = [...restTimes];
  if (arr.length < MAX_STANDARD_REST_TIME_COUNT) {
    arr.push(TimeRange.create(value));
  }
  return arr;
};
