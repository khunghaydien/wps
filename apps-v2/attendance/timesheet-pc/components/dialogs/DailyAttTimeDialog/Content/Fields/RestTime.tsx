import * as React from 'react';

import styled from 'styled-components';

import $AttTimeRangeField from '@apps/commons/components/fields/AttTimeRangeField';
import msg from '@apps/commons/languages';

import { RestTimeReason } from '@attendance/domain/models/RestTimeReason';

import RestReason from './RestReason';
import * as helper from '@attendance/ui/helpers/dailyRecord';
import ImgBtnMinusField from '@attendance/timesheet-pc/images/btnMinusField.png';
import ImgBtnPlusField from '@attendance/timesheet-pc/images/btnPlusField.png';

// FIXME: https://github.com/microsoft/TypeScript/issues/37597
// @ts-ignore
const IconButton = styled.button.attrs(() => ({
  type: 'button',
}))`
  padding: 0;
  border: 0;
  margin: 0;
  background-color: transparent;
  cursor: pointer;
  :hover > img,
  :focus > img {
    filter: grayscale(50%);
  }
`;

const AddIconButton: React.FC<{ onClick: () => void }> = ({ onClick }) => (
  <IconButton onClick={onClick}>
    <img src={ImgBtnPlusField} alt={msg().Att_Btn_AddItem} />
  </IconButton>
);

const DeleteIconButton: React.FC<{ onClick: () => void }> = ({ onClick }) => (
  <IconButton onClick={onClick}>
    <img src={ImgBtnMinusField} alt={msg().Att_Btn_RemoveItem} />
  </IconButton>
);

const RestTimeItemContainer = styled.div`
  display: flex;
`;

const RestTimeItemIconCell = styled.div`
  display: flex;
  justify-content: center;
  align-items: center;
  width: 30px;
  margin-right: 7px;
  :last-child: {
    margin-right: 0px;
  }
`;

const RestReasonCell = styled.div`
  width: 250px;
  margin-right: 15px;
`;

const AttTimeRangeField = styled($AttTimeRangeField)`
  margin-right: 15px;

  // FIXME: 下記がないと、ボタンとの間のスペースが意図しない長さになる
  .commons-fields-att-time-range-field__separation {
    padding: 5px;
  }
`;

const RestTime: React.FC<{
  record: {
    startTime: number | null;
    endTime: number | null;
    restReason: RestTimeReason | null;
  };
  restTimeReasons: RestTimeReason[];
  enabledRestReason: boolean;
  enableAdd: boolean;
  enableDelete: boolean;
  readOnly?: boolean;
  onUpdateStart: (value: number | null) => void;
  onUpdateEnd: (value: number | null) => void;
  onUpdateReason: (value: RestTimeReason | null) => void;
  onClickAdd: () => void;
  onClickDelete: () => void;
}> = ({
  record,
  restTimeReasons,
  enabledRestReason,
  enableAdd,
  enableDelete,
  readOnly,
  onUpdateStart,
  onUpdateEnd,
  onUpdateReason: $onUpdateReason,
  onClickAdd,
  onClickDelete,
}) => {
  const onUpdateReason = React.useCallback(
    (value: RestTimeReason | null) => {
      $onUpdateReason(value);
    },
    [$onUpdateReason]
  );
  return (
    <RestTimeItemContainer>
      <AttTimeRangeField
        startTime={helper.time.toHHmm(record.startTime)}
        endTime={helper.time.toHHmm(record.endTime)}
        onBlurAtStart={(value: string) =>
          onUpdateStart(helper.time.toNumberOrNull(value))
        }
        onBlurAtEnd={(value: string) =>
          onUpdateEnd(helper.time.toNumberOrNull(value))
        }
        disabled={readOnly}
      />
      {enabledRestReason && (
        <RestReasonCell>
          <RestReason
            value={record.restReason}
            restTimeReasons={restTimeReasons}
            onUpdateReason={onUpdateReason}
            readOnly={readOnly}
          />
        </RestReasonCell>
      )}
      <RestTimeItemIconCell>
        {enableAdd && !readOnly ? <AddIconButton onClick={onClickAdd} /> : null}
      </RestTimeItemIconCell>
      <RestTimeItemIconCell>
        {enableDelete && !readOnly ? (
          <DeleteIconButton onClick={onClickDelete} />
        ) : null}
      </RestTimeItemIconCell>
    </RestTimeItemContainer>
  );
};
RestTime.defaultProps = {
  readOnly: false,
};

export default RestTime;
