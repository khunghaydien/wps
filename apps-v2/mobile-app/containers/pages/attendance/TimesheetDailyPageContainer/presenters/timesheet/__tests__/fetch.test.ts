import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';

import { MAX_STANDARD_REST_TIME_COUNT } from '@apps/attendance/domain/models/RestTime';

import presenter from '../fetch';
import {
  IInputData,
  IOutputData,
} from '@attendance/domain/useCases/timesheet/IFetchEntityUseCase';

jest.mock('uuid/v4');

const mockStore = configureMockStore([thunk]);

describe('complete()', () => {
  it('should do.', () => {
    // Arrange
    const store = mockStore({});

    // Act
    presenter(store)(undefined as unknown as IInputData).complete({
      employeeId: 'employeeId',
      timesheet: {
        workingType: {
          useObjectivelyEventLog: true,
        },
      },
      maxRestTimesCount: MAX_STANDARD_REST_TIME_COUNT,
    } as unknown as IOutputData);

    // Assert
    expect(store.getActions()).toMatchSnapshot();
  });
  it('should do with isMigratedSummary.', () => {
    // Arrange
    const store = mockStore({});

    // Act
    presenter(store)(undefined as unknown as IInputData).complete({
      employeeId: 'employeeId',
      timesheet: {
        isMigratedSummary: true,
      },
      dailyObjectivelyEventLogs: 'dailyObjectivelyEventLogs',
    } as unknown as IOutputData);

    // Assert
    expect(store.getActions()).toMatchSnapshot();
  });
});
describe('error()', () => {
  it('should do catchApiError if error is not occurred', async () => {
    // Arrange
    const store = mockStore({
      common: {
        app: {
          error: null,
        },
      },
    });

    // Act
    presenter(store)(undefined as unknown as IInputData).error({
      message: 'Error Test',
    });

    // Assert
    expect(store.getActions()).toMatchSnapshot();
  });

  it('should not do catchApiError if error is occurred', async () => {
    // Arrange
    const store = mockStore({
      common: {
        app: {
          error: 'error',
        },
      },
    });

    // Act
    presenter(store)(undefined as unknown as IInputData).error({
      message: 'Error Test',
    });

    // Assert
    expect(store.getActions()).toMatchSnapshot();
  });
});

describe('start()', () => {
  it('should do.', () => {
    // Arrange
    const store = mockStore({});

    // Act
    presenter(store)(undefined as unknown as IInputData).start();

    // Assert
    expect(store.getActions()).toMatchSnapshot();
  });
});

describe('finally()', () => {
  it('should do.', () => {
    // Arrange
    const store = mockStore({});

    // Act
    presenter(store)(undefined as unknown as IInputData).finally();

    // Assert
    expect(store.getActions()).toMatchSnapshot();
  });
});
