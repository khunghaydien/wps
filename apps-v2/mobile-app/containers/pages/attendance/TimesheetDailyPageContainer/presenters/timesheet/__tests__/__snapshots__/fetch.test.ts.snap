// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`complete() should do with isMigratedSummary. 1`] = `
Array [
  Object {
    "payload": Object {
      "targetDate": undefined,
      "timesheet": Object {
        "isMigratedSummary": true,
      },
    },
    "type": "MOBILE-APP/ATTENDANCE/TIMESHEET/UI/DAILY/EDITING/FETCH_SUCCESS",
  },
  Object {
    "payload": Object {
      "targetDate": undefined,
      "timesheet": Object {
        "isMigratedSummary": true,
      },
    },
    "type": "MOBILE-APP/ATTENDANCE/TIMESHEET/UI/DAILY/PAGING/FETCH_SUCCESS",
  },
  Object {
    "payload": Object {
      "isMigratedSummary": true,
    },
    "type": "MOBILE_APP/MODULES/ATTENDANCE/TIMESHEET/ENTITIES/FETCH_SUCCESS",
  },
]
`;

exports[`complete() should do. 1`] = `
Array [
  Object {
    "payload": Object {
      "targetDate": undefined,
      "timesheet": Object {
        "workingType": Object {
          "useObjectivelyEventLog": true,
        },
      },
    },
    "type": "MOBILE-APP/ATTENDANCE/TIMESHEET/UI/DAILY/EDITING/FETCH_SUCCESS",
  },
  Object {
    "payload": Object {
      "targetDate": undefined,
      "timesheet": Object {
        "workingType": Object {
          "useObjectivelyEventLog": true,
        },
      },
    },
    "type": "MOBILE-APP/ATTENDANCE/TIMESHEET/UI/DAILY/PAGING/FETCH_SUCCESS",
  },
  Object {
    "payload": Object {
      "workingType": Object {
        "useObjectivelyEventLog": true,
      },
    },
    "type": "MOBILE_APP/MODULES/ATTENDANCE/TIMESHEET/ENTITIES/FETCH_SUCCESS",
  },
]
`;

exports[`error() should do catchApiError if error is not occurred 1`] = `
Array [
  Object {
    "payload": Object {
      "errorCode": undefined,
      "isContinuable": true,
      "isFunctionCantUseError": false,
      "message": "APIエラー - Error Test",
      "problem": "Error Test",
      "stackTrace": "",
      "type": "APIエラー",
    },
    "type": "CATCH_API_ERROR",
  },
  Object {
    "type": "MOBILE-APP/ATTENDANCE/TIMESHEET/UI/DAILY/PAGING/CLEAR",
  },
  Object {
    "type": "MOBILE-APP/ATTENDANCE/TIMESHEET/UI/DAILY/EDITING/CLEAR",
  },
  Object {
    "type": "MOBILE-APP/ATTENDANCE/TIMESHEET/UI/DAILY/REST_REASONS/CLEAR",
  },
  Object {
    "type": "MOBILE_APP/MODULES/ATTENDANCE/TIMESHEET/ENTITIES/CLEAR",
  },
]
`;

exports[`error() should not do catchApiError if error is occurred 1`] = `
Array [
  Object {
    "payload": Object {
      "errorCode": undefined,
      "isContinuable": true,
      "isFunctionCantUseError": false,
      "message": "APIエラー - Error Test",
      "problem": "Error Test",
      "stackTrace": "",
      "type": "APIエラー",
    },
    "type": "CATCH_API_ERROR",
  },
  Object {
    "type": "MOBILE-APP/ATTENDANCE/TIMESHEET/UI/DAILY/PAGING/CLEAR",
  },
  Object {
    "type": "MOBILE-APP/ATTENDANCE/TIMESHEET/UI/DAILY/EDITING/CLEAR",
  },
  Object {
    "type": "MOBILE-APP/ATTENDANCE/TIMESHEET/UI/DAILY/REST_REASONS/CLEAR",
  },
  Object {
    "type": "MOBILE_APP/MODULES/ATTENDANCE/TIMESHEET/ENTITIES/CLEAR",
  },
]
`;

exports[`finally() should do. 1`] = `
Array [
  Object {
    "payload": undefined,
    "type": "END_LOADING",
  },
]
`;

exports[`start() should do. 1`] = `
Array [
  Object {
    "payload": undefined,
    "type": "START_LOADING",
  },
]
`;
